src/Dictionaries/ExecutablePoolDictionarySource.cpp

```diff
@@ -0,0 +1,272 @@
+{
+    std::unique_ptr<ShellCommand> process;
+    process_pool->pop(process);
+
+    size_t rows_to_read = block.rows();
+    auto read_stream = context.getInputFormat(format, process->out, sample_block, rows_to_read);
+
+    auto stream = std::make_unique<PoolBlockInputStreamWithBackgroundThread>(
+        process_pool, std::move(read_stream), std::move(process), rows_to_read,
+        [block, this](WriteBufferFromFile & out) mutable
+        {
+            auto output_stream = context.getOutputStream(format, out, block.cloneEmpty());
+            formatBlock(output_stream, block);
+            std::cerr << "Write block to process " << std::endl;
+        });
+
+    if (implicit_key)
+        return std::make_shared<BlockInputStreamWithAdditionalColumns>(block, std::move(stream));
+    else
+        return std::shared_ptr<PoolBlockInputStreamWithBackgroundThread>(stream.release());
+}
+
+bool ExecutablePoolDictionarySource::isModified() const
+{
+    return true;
+}
+
+bool ExecutablePoolDictionarySource::supportsSelectiveLoad() const
+{
+    return true;
+}
+
+bool ExecutablePoolDictionarySource::hasUpdateField() const
+{
+    return !update_field.empty();
+}
+
+DictionarySourcePtr ExecutablePoolDictionarySource::clone() const
+{
+    return std::make_unique<ExecutablePoolDictionarySource>(*this);
+}
+
+std::string ExecutablePoolDictionarySource::toString() const
+{
+    return "ExecutablePool size: " + std::to_string(pool_size) + " command: " + command;
+}
+
+void registerDictionarySourceExecutablePool(DictionarySourceFactory & factory)
+{
+    auto create_table_source = [=](const DictionaryStructure & dict_struct,
+                                 const Poco::Util::AbstractConfiguration & config,
+                                 const std::string & config_prefix,
+                                 Block & sample_block,
+                                 const Context & context,
+                                 const std::string & /* default_database */,
+                                 bool check_config) -> DictionarySourcePtr
+    {
+        if (dict_struct.has_expressions)
+            throw Exception{"Dictionary source of type `executable_pool` does not support attribute expressions", ErrorCodes::LOGICAL_ERROR};
+
+        /// Executable dictionaries may execute arbitrary commands.
+        /// It's OK for dictionaries created by administrator from xml-file, but
+        /// maybe dangerous for dictionaries created from DDL-queries.
+        if (check_config)
+            throw Exception("Dictionaries with executable pool dictionary source is not allowed created from DDL are not allowed", ErrorCodes::DICTIONARY_ACCESS_DENIED);
+
+        Context context_local_copy = copyContextAndApplySettings(config_prefix, context, config);
+
+        auto settings_no_parallel_parsing = context_local_copy.getSettings();
+        settings_no_parallel_parsing.input_format_parallel_parsing = false;
```

Comment:
Why? Is there any bug with parallel parsing?

Reply:
Currently parallel parsing input format cannot read exactly max_block_size rows from input, so it will be blocked on ReadBufferFromFileDescriptor because this file descriptor represent pipe that does not have eof.